### ISSUE TITLE
Problem with large regression numbers

### ISSUE THREAD

**Issue by kylelee82** [Microsoft/LightGBM#1278]
## Environment info
Operating System: Ubuntu 16.04
CPU: Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz
C++/Python/R version: Python version==2.1.0

## Problem:

It seems like for regression targets larger than 1million, there is a performance issue with LightGBM and the model doesn't really train i.e. bad convergence.  It is surprising that targets of million/billion (which are pretty decently sized regression targets) already hit the limit...

## Reproducible examples

Please refer to attached file `debug_lgbm.py.gz` for example.  This can be run directly.

[debug_lgbm.py.gz](https://github.com/Microsoft/LightGBM/files/1832197/debug_lgbm.py.gz)

`SCALE_LIST` refers to the range of scaling applied on the target (i.e. `Y_target = Y / scale`).  The prediction is then rescaled (multiplied) by by the scaling above to compute the final RMSE.

Since the targets of the dataset above is already normalized to a unit scale (max = 5, min=0.15), a scale of 1e-5 would give a range of <1million, 1e-8 is < 1 billion, etc.

You can change `MODE` to `xgb` for another popular Gradient Boosting example (which works OK in the large number regime, but actually worse off than LightGBM for small targets - but that's another issue).

## Graphs

In both graphs below, the stable RMSE range is around ~ 0.45-0.5

**RMSE vs Target scaling factor for LightGBM**

![image](https://user-images.githubusercontent.com/7127211/37696210-f76423be-2d0f-11e8-9c7c-0b0314b24ce4.png)

You'll note that the RMSE jump occurs at a scaling of 1e-5 to 1e-6, which would transform the target to ~1 million range.

**RMSE vs Target scaling factor for the other popular Gradient Boosting framework**

![image](https://user-images.githubusercontent.com/7127211/37696530-232e64bc-2d12-11e8-9cc4-db7bb010aee6.png)

No issues for this framework at large numbers but it gets worse at smaller numbers, possibly because of significant digits during evaluation (but that's another topic)

**Comment by guolinke** [comment]
please try the “reg_sqrt=true”

BTW, Lightgbm will limit the max output of tree leaves(default is 100), you can change it to a larger value.

**Comment by kylelee82** [comment]
> please try the “reg_sqrt=true”

Thanks a lot - this will help, see updated graph below:

![image](https://user-images.githubusercontent.com/7127211/37698654-c3615fbe-2d1e-11e8-8506-14c41a8f359d.png)

But I have some questions on doing this:

1. Is this recommended for all regression cases? (i.e. just leave it True for all regression cases)
2. Will the log-link property for some of the objectives still hold if this transformation is used?
3. I assume it will not properly work for negative numbers?

**Comment by chivee** [comment]
Is this recommended for all regression cases? (i.e. just leave it True for all regression cases)

>  it only can be used in the regression case, however, it recommend when your regression target was too lager and sparse. it will make the target range more 'compactful'.

Will the log-link property for some of the objectives still hold if this transformation is used?

> I'm not quite sure what is log-link property, but from our experience, log transformation has poor performance then sqrt transformation. 

I assume it will not properly work for negative numbers?

> actually, it will works.

**Comment by kylelee82** [comment]
> I'm not quite sure what is log-link property, but from our experience, log transformation has poor performance then sqrt transformation.

For clarification I was referring to the exotic loss functions (Poison/Tweedie/Gamma), and it seems this feature is blocked out when those functions are used:

https://github.com/Microsoft/LightGBM/blob/master/src/objective/regression_objective.hpp#L368

**Comment by guolinke** [comment]
@kylelee82 
it depends on the distribution of the labels. if it is a Poisson/Tweedie/Gamm distribution, you can use it.

`reg_sqrt` cannot be used here since it will break the distribution assumption of Poisson/Tweedie/Gamm .

**Comment by kylelee82** [comment]
@guolinke So moving forward, the official usage for behavior with large numbers in regression (>million range) will just to enable `reg_sqrt` and no further improvements will be made in the engine itself, right?

**Comment by guolinke** [comment]
yeah.
